zebra-rpc/src/methods/get_block_template_rpcs.rs

```diff
@@ -333,40 +333,113 @@ where
 
+            let mut client_long_poll_id = None;
             if let Some(parameters) = parameters {
-                check_block_template_parameters(parameters)?;
+                check_block_template_parameters(&parameters)?;
+
+                client_long_poll_id = parameters.long_poll_id;
             }
 
             // - Checks and fetches that can change during long polling
-
-            // Check if we are synced to the tip.
-            // The result of this check can change during long polling.
-            check_synced_to_tip(network, latest_chain_tip, sync_status)?;
-
-            // Fetch the state data and local time for the block template:
-            // - if the tip block hash changes, we must return from long polling,
-            // - if the local clock changes on testnet, we might return from long polling
             //
-            // We also return after 90 minutes on mainnet, even if we have the same response.
-            let chain_tip_and_local_time = fetch_state_tip_and_local_time(state).await?;
+            // Set up the loop.
+            let mut max_time_reached = false;
+
+            // The loop returns the server long poll ID,
+            // which should be different to the client long poll ID.
+            let (server_long_poll_id, chain_tip_and_local_time, mempool_txs) = loop {
+                // Check if we are synced to the tip.
+                // The result of this check can change during long polling.
+                //
+                // TODO:
+                // - add `async changed()` methods to ChainTip and ChainSyncStatus
+                //   (using `changed() -> Changed` and `impl Future<()> for Changed`)
+                check_synced_to_tip(network, latest_chain_tip.clone(), sync_status.clone())?;
+
+                // Fetch the state data and local time for the block template:
+                // - if the tip block hash changes, we must return from long polling,
+                // - if the local clock changes on testnet, we might return from long polling
+                //
+                // We always return after 90 minutes on mainnet, even if we have the same response,
+                // because the max time has been reached.
+                //
+                // TODO: timeout and exit the loop when max time is reached
+                let chain_tip_and_local_time =
+                    fetch_state_tip_and_local_time(state.clone()).await?;
 
-            // Fetch the mempool data for the block template:
-            // - if the mempool transactions change, we might return from long polling.
-            let mempool_txs = fetch_mempool_transactions(mempool).await?;
+                // Fetch the mempool data for the block template:
+                // - if the mempool transactions change, we might return from long polling.
+                //
+                // TODO:
+                // - add a `MempoolChange` type with an `async changed()` method.
```

Comment:
We might be able to use `mempool_transaction_receiver` here.

Reply:
Using watch channels is out of scope for this PR, see the TODOs.